### PR TITLE
handle completely unscanned storages in the background scanner

### LINKED ISSUE #13876: Folder Size does not get calculated
Ubuntu 14.04, nginx, MySQL, PHP5.5.9, ownCloud git/Master 8.0.2

I created a new dev instance and added one SMB mount point (files external) successfully.

Problem:
when I go to the Files App, the size of the added external mount point never gets calculated, even not after a day of waiting time.
Logout/relogin, rebooting ect does not change the status. I always get "Ausstehend".

I can access the folder and its contents, I even can use gallery, but the Files App keeps staying on "Ausstehend".

There are no log entries

Info. This worked before on 8.x and works on 7.0.4 stable

### PR DESCRIPTION
**Pull request by icewind1991**
Fixes #13876

previously if there were no cache entries for a storage (folder containing the storage has never been visited after the storage has been added) the background scanner is not triggered, this fixes it by triggering a full scan if the storage is unscanned.

cc @PVince81 @mmattel

### CODE REVIEW

File: lib/private/files/cache/scanner.php [hunk 1]
```diff
@@ -448,26 +448,30 @@ public static function isPartialFile($file) {
...
 	public function backgroundScan() {
-		$lastPath = null;
-		while (($path = $this->cache->getIncomplete()) !== false && $path !== $lastPath) {
-			try {
-				$this->scan($path, self::SCAN_RECURSIVE, self::REUSE_ETAG);
-				\OC_Hook::emit('Scanner', 'correctFolderSize', array('path' => $path));
-				if ($this->cacheActive) {
-					$this->cache->correctFolderSize($path);
+		if (!$this->cache->inCache('')) {
+			$this->scan('', self::SCAN_RECURSIVE, self::REUSE_ETAG);
```
Comment (PVince81) [inline lib/private/files/cache/scanner.php]: We should also catch the exceptions here.

Comment (icewind1991) [inline lib/private/files/cache/scanner.php]: done

File: lib/private/files/cache/scanner.php [hunk 2]
```diff
@@ -465,9 +460,28 @@ public function backgroundScan() {
...
 			}
-			// FIXME: this won't proceed with the next item, needs revamping of getIncomplete()
-			// to make this possible
-			$lastPath = $path;
+		} else {
+			$lastPath = null;
+			while (($path = $this->cache->getIncomplete()) !== false && $path !== $lastPath) {
+				try {
+					$this->scan($path, self::SCAN_RECURSIVE, self::REUSE_ETAG);
+					\OC_Hook::emit('Scanner', 'correctFolderSize', array('path' => $path));
```
Comment (PVince81) [inline lib/private/files/cache/scanner.php]: and what about this hook ?

It rather looks like the "try" block should be moved to a function and called from both places ?

Comment (PVince81) [inline lib/private/files/cache/scanner.php]: Another quicker solution: only scan the root, non-recursive, to auto-create the root entry with size -1. And then proceed with the regular `while`

Comment (icewind1991) [inline lib/private/files/cache/scanner.php]: done

Comment (icewind1991) [inline lib/private/files/cache/scanner.php]: > Another quicker solution: only scan the root, non-recursive, to auto-create the root entry with size -1. And then proceed with the regular while

A bit more ugly imo and will result in the same thing (the entire thing being scanned)

**Comment by PVince81** [comment]
Second review @mmattel @Xenopathic @nickvergessen @blizzz ?

@karlitschek backport to stable9 ?

**Comment by PVince81** [comment]
@icewind1991 please fix the unit test
```
Test\Files\Cache\Scanner::testBackgroundScan
Undefined variable: path

/ssd/jenkins/workspace/core-ci-linux/database/sqlite/label/SLAVE/lib/private/files/cache/scanner.php:471
/ssd/jenkins/workspace/core-ci-linux/database/sqlite/label/SLAVE/lib/private/files/cache/scanner.php:460
/ssd/jenkins/workspace/core-ci-linux/database/sqlite/label/SLAVE/tests/lib/files/cache/scanner.php:142
```

**Comment by karlitschek** [comment]
great. please backport

**Comment by icewind1991** [comment]
> @icewind1991 please fix the unit test

Fixed

**Comment by icewind1991** [comment]
stable9: https://github.com/owncloud/core/pull/23576

**Comment by mmattel** [comment]
@icewind1991 

- Rebased my testinstance and applied this PR
- Added a big enough new and never before used SMB share (green light at the admin page)
- Back in files it shows "Ausstehend" (Pending)
- Monitoring via htop processor activities - more or less nothing during the next 15min
- Monitoring oc_filecache table filtering the corresponding storage ID, no changes except one entry (parent = -1)
- Double check if the PR has really been applied = ok
- Nothing in the log

**Comment by mmattel** [comment]
- No changes after log off an relogon
- No changes now after about 20min
- No changes even admin page tells, last cron run 14 min ago respectively just now

UPDATE:
After now about 30 min (can not determine exactly), the scanner started (htop showed load), oc_filecache has been filled and on the files page, the new SMB storage shows the size.
The PR is working.

Question: There is a long time between adding a storage and starting the background scan.
I excpected a delay of less than 5min. And the message on the admin page that the last cron run but nothing hapened was very confusing in this situation.

**Comment by PVince81** [comment]
@mmattel good point... the reason it starts late is because of the time where cron runs.
Not sure how to solve this, maybe automatically reset that job's last_run ?
Or add a new temp bkg job after creating the storage ?

@icewind1991

**Comment by mmattel** [comment]
Just thinking around:

1. If you create an external storage, wait for 3-5min (tbd) to start the background job. Trigger for the timer is the green light = storage is successfully attachable.
2. If in the meanwhile of this waiting time something has changed on the storage settings, restart the timer.
3. If you delete the storage during the waiting time, delete the background job.
4. If you change to files AND open on the newly created storage which has not been scanned, start scanning immediately = set the delay timer to zero.

In the current implementation, what happens on 3?
Does the backgrund job start on an non existing storage?

**Comment by icewind1991** [comment]
> In the current implementation, what happens on 3?
Does the backgrund job start on an non existing storage?

The background job is not storage specific so it will be handled just fine.

For now I would just go with the current approach since it's only for showing the folder size in the web ui, the external storage will always be accessible no matter when or if the background job fires

**Pull request merged**